src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,106 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] arrayList;
+    private int size;
+    private int capacity;
+
     @Override
-    public void add(T value) {
+    protected Object clone() throws CloneNotSupportedException {
+        return super.clone();
+    }
 
+    public void growCapacity() {
+        if (isEmpty()) {
+            arrayList = (T[]) new Object[DEFAULT_CAPACITY];
+            capacity = DEFAULT_CAPACITY;
+        }
+        T[] initialArray = arrayList;
+        arrayList = (T[]) new Object[capacity + capacity / 2];
+        System.arraycopy(initialArray, 0, arrayList, 0, size);
+        capacity = arrayList.length;
     }
 
     @Override
-    public void add(T value, int index) {
+    public void add(T value) {
+        if (size == capacity) {
+            growCapacity();
+        }
+        arrayList[size++] = value;
+    }
 
+    @Override
+    public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of range: " + index);
+        }
+        if (size == capacity) {
+            growCapacity();
+        }
+        if (index == size) {
+            add(value);
+        } else {
+            System.arraycopy(arrayList, index, arrayList, index + 1, size - index);
+            arrayList[index] = value;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index);
+        return arrayList[index];
```

Comment:
```suggestion
        return (T) arrayList[index];
```

Reply:
Please, explain why type casting is necessary? As I understand, the element has type T.

Reply:
You are right, my mistake, didn't notice that you create T[] elements.